Improve pipelines docs (#7012)
mlflow/mlflow: docs/source/pipelines.rst

MLflow Pipelines (experimental)
===============================

MLflow Pipelines is a framework that enables you to quickly develop high-quality models and deploy
them to production. Compared to ad-hoc ML workflows, MLflow Pipelines offers several major benefits:

> - **Pipeline templates**: :ref:`Predefined templates <pipeline-templates>` for common ML tasks,

Reviewer: I would say the benefits directly instead of the components behind them:
* Quicker to get started.
* Faster to iterate during model development.
* Easier to ship to production.

(Wording doesn't need to be exact.)

  such as :ref:`regression modeling <regression-template>`, enable you to get started quickly and
  focus on building great models, eliminating the large amount of boilerplate code that is
  traditionally required to curate datasets, engineer features, train & tune models, and package
  models for production deployment.

- **Pipeline engine**: The intelligent pipeline execution engine accelerates model development by
  caching results from each step of the process and re-running the minimal set of steps as changes
  are made.

- **Production-ready structure**: The modular, git-integrated :ref:`pipeline structure
  <pipeline-templates-key-concept>` dramatically simplifies the handoff from development to
  production by ensuring that all model code, data, and configurations are easily reviewable and
  deployable by ML engineers.

Quickstarts
-----------

Prerequisites
~~~~~~~~~~~~~

.. _pipelines-installation:

MLflow Pipelines is available as an extension of the
`MLflow Python library <https://pypi.org/project/mlflow/>`_. You can install MLflow Pipelines
as follows:

> - **Windows and Unix systems**: Install MLflow Pipelines from PyPI:

Reviewer: minor: not sure Mac users know that Mac is a unix system. Shall we just say `Local` and `Databricks`? They also match the profile names in our example.

  .. code-block:: sh

      pip install mlflow[pipelines]

  Note that MLflow Pipelines requires Make, which may not be preinstalled on some Windows systems.

>   Windows users must install Make before using MLflow Pipelines.

Reviewer: Shall we provide a link to show how to install make on Windows?

- **Databricks**: Install MLflow Pipelines from a Databricks Notebook by running
  ``pip install mlflow[pipelines]``, or install MLflow Pipelines on a Databricks Cluster by
  following the PyPI library installation instructions at
  https://docs.databricks.com/libraries/cluster-libraries.html#install-a-library-on-a-cluster
  and specifying the ``mlflow[pipelines]`` package string. Note that
  `Databricks Runtime <https://docs.databricks.com/runtime/dbr.html>`_ version 11.0
  or greater is required in order to install MLflow Pipelines on Databricks.

NYC taxi fare prediction example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The `NYC taxi fare prediction example <https://github.com/mlflow/mlp-regression-example>`_

> uses the |MLflow Pipelines Regression Template| to develop and score accurate models on the

Reviewer: nit: what does "accurate" mean here?

`NYC Taxi (TLC) Trip Record Dataset
<https://www1.nyc.gov/site/tlc/about/tlc-trip-record-data.page>`_. You can run the example by
:ref:`installing MLflow Pipelines <pipelines-installation>` and running the `Jupyter example
notebook <https://github.com/mlflow/mlp-regression-example/blob/master/notebooks/jupyter.ipynb>`_

> or the `Databricks example notebook

Reviewer: "or" -> "locally or ... on Databricks"

Author: Fixed!

<https://github.com/mlflow/mlp-regression-example/blob/master/notebooks/databricks.py>`_. For more
information about importing and running a Databricks notebook, see

> https://docs.databricks.com/notebooks/notebooks-manage.html#import-a-notebook.

Reviewer: We should provide the repos instructions here (https://docs.databricks.com/repos/sync-remote-repo.html). If a user just import the notebook, it won't work.

Author: Fixed - sorry about that :)

> You can also modify the example to build and score a model on a dataset of your choosing.

Reviewer: Do we want to recommend using the template instead of the example for your own projects? It is not very clear to me.

Author: Changed this to recommend using the template.

For more information about modifying the |MLflow Pipelines Regression Template|, see the
|Regression Template reference guide|.

Key concepts
------------

.. _steps-key-concept:

> - **Steps**: Steps represent individual ML operations, such as ingesting data, fitting an estimator,

Reviewer 2: ```suggestion
- **Steps**: A step represents an individual ML operation, such as ingesting data, fitting an estimator,
```

Author: Done!

>   evaluating a model against test data, or deploying a model for real-time scoring. Steps accept a

Reviewer 2: ```suggestion
  evaluating a model against test data, or deploying a model for real-time scoring. Each step accepts a
```

Author: Done!

  collection of well-defined inputs and produce well-defined outputs according to user-defined
  configurations and code.

.. _pipelines-key-concept:

> - **Pipelines**: A Pipeline is a composition of :ref:`Steps <steps-key-concept>` used to solve an

Reviewer 2: ```suggestion
- **Pipelines**: A Pipeline is an ordered composition of :ref:`Steps <steps-key-concept>` used to solve an
```

Author: Good idea! Done!

  ML problem or perform an MLOps task, such as developing a regression model or performing batch
  model scoring on production data. MLflow Pipelines provides
  :py:class:`APIs <mlflow.pipelines.Pipeline>` and a :ref:`CLI <cli>` for running pipelines and
  inspecting their results.

.. _pipeline-templates-key-concept:

- **Templates**: A Pipeline Template is a git repository with a :ref:`standardized, modular layout

>   <pipeline-template-structure>` containing all of the customizable code and configurations for a

Reviewer 2: ```suggestion
  <pipeline-template-structure>` containing all of the customizable code and configurations required for a
```

Author: I don't think the word `required` is necessary here.

Reviewer 2: I feel like something is missing gramatically, if you feel `required` is too prescriptive then maybe `code and configurations used by a ...`  or something?

  :ref:`Pipeline <pipelines-key-concept>`. Each template also defines its requirements, data science
  notebooks, and tests. MLflow Pipelines includes :ref:`predefined templates <pipeline-templates>`
  for a variety of model development and MLOps tasks.

.. _profiles-key-concept:

> - **Profiles**: Profiles contain user-specific or environment-specific configurations for a

Reviewer: We haven't mentioned `pipeline.yaml` yet. Maybe we can say `pipeline.yam and profiles` here.

Author: I don't think `pipeline.yaml` is a separate concept. It's part of the template concept above. I've added `pipeline.yaml` to the template concept description; let me know if that works

  :ref:`Pipeline <pipelines-key-concept>`, such as the particular set of hyperparameters being
  tuned by a data scientist in development or the MLflow Model Registry URI and credentials
  used to store production-worthy models. Each profile is represented as a YAML file in the
  :ref:`Pipeline Template <pipeline-templates-key-concept>`.

.. _step-cards-key-concept:

> - **Step Cards**: Step Cards display :ref:`Pipeline <pipelines-key-concept>` results, including

Reviewer 2: I think this is potentially confusing - they display step results, not those of the whole Pipeline, which is a difference more acute for the earlier steps like ingestion.

Author: Good catch! I've adjusted this. Let me know what you think!

  dataset profiles, model performance & explainability plots, overviews of the best model
  parameters found during tuning, and more. Step Cards and their corresponding dataset and model
  information are also logged to MLflow Tracking.

> Usage

Author: @jinzhang21 This section is unchanged, as usage updates are assigned to you via ML-24619. Feel free to push changes to this PR or file a separate PR.

-----

The general workflow for using MLflow Pipelines is as follows:

1. Clone a :ref:`Pipeline Template <pipeline-templates>` git repository corresponding to the ML
   problem that you want to solve or the MLOps task that you want to perform. View the template's
   README file for information about the :ref:`Pipeline Steps <steps-key-concept>` that it defines
   and the results that it produces.

    .. code-block:: sh
      :caption: An example of cloning the |MLflow Pipelines Regression Template|

      git clone https://github.com/mlflow/mlp-regression-template

    .. note::
      On Databricks, we recommend cloning the :ref:`Pipeline Template <pipeline-templates>` git

>       repository using |Databricks Repos|.

Reviewer 2: Isn't repos their only option?

Author: You could technically do `%sh git clone ...`, which is what we do for integration tests. Since this section is preexisting and updates are owned by @jinzhang21 , I'm going to leave this as-is for now.

Reviewer 2: Don't have all the context, but that seems a little contrived unless a lot of people are doing it that way. Could we be prescriptive instead? ie `On Databricks, clone the ... git repository using Databricks Repos`

Reviewer 2: Don't have all the context, but that seems a little contrived unless a lot of people are doing it that way. Could we be prescriptive instead? ie `On Databricks, clone the ... git repository using Databricks Repos`

Reviewer 3: Makes sense to me. I'll update this section later.

    .. note::
      By default, MLflow Pipelines caches results from each :ref:`Pipeline Step
      <steps-key-concept>` within the ``.mlflow`` subdirectory of the home folder on the
      local filesystem. The ``MLFLOW_PIPELINES_EXECUTION_DIRECTORY`` environment variable can
      be used to specify an alternative location for caching results.

.. _pipeline-templates:

Pipeline Templates
------------------

MLflow Pipelines currently offers the following predefined templates that can be easily customized
to develop and deploy high-quality, production-ready models for your use cases:

.. _regression-template:

- **MLflow Pipelines Regression Template**: The `MLflow Pipelines Regression Template
  <https://github.com/mlflow/mlp-regression-template>`_ is designed for developing and scoring
  regression models. For more information, see the |Regression Template reference guide| and
  the :ref:`Regression Pipeline API documentation <mlflow-regression-pipeline>`.

Additional pipelines for a variety of ML problems and MLOps tasks are under active development.

Detailed reference guide
------------------------

.. _pipeline-template-structure:

Template structure
~~~~~~~~~~~~~~~~~~

> Pipeline Templates are git repositories with a standardized, modular layout. The following

Reviewer 2: What does modular mean here? The pipeline is modularized but I'm not sure how that term applies to the bare file structure.

Author: Good question! It means that the pipeline definition is decomposed into individual steps that each have their own code confs and code files, in contrast to a notebook that globs together all of the pipeline content in the same place and is hard to review / productionize.

MLflow Pipelines (experimental)
===============================

MLflow Pipelines is a framework that enables data scientists to quickly develop high-quality models
and deploy them to production. Compared to ad-hoc ML workflows, MLflow Pipelines offers several
major benefits:

> - **Get started quickly**:ref:`Predefined templates <pipeline-templates>` for common ML tasks,

Reviewer 4: ```suggestion
- **Get started quickly**: :ref:`Predefined templates <pipeline-templates>` for common ML tasks,
```
